In [1]:
import numpy as np
import pandas as pd
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.stats.proportion import proportions_ztest, proportion_effectsize
from statsmodels.stats.power import NormalIndPower

np.random.seed(42)

# Task 1: Chi-Square Goodness-of-Fit Test


In [2]:
#step 1
observed = np.array([120, 85, 110, 85])
observed

array([120,  85, 110,  85])

In [3]:
#step 2
total_players = observed.sum()
expected = np.array([total_players * 0.25] * 4)
expected

array([100., 100., 100., 100.])

In [4]:
#step 3
chi2_stat, p_value = st.chisquare(observed, f_exp=expected)
chi2_stat, p_value

(np.float64(9.5), np.float64(0.023331360430831553))

In [5]:
# step 4
df = len(observed) - 1

print(f"Chi-square statistic: {chi2_stat:.2f}")
print(f"Degrees of freedom: {df}")
print(f"P-value: {p_value:.4f}")

Chi-square statistic: 9.50
Degrees of freedom: 3
P-value: 0.0233


In [6]:
#step 5 
if p_value < 0.05:
    print("Reject H0")
else:
    print("Fail to reject H0")

Reject H0


### step 6 
Compared with the expected count of 100 players per class, Warrior is over-represented by 20 players and Rogue by 10 players, while Mage and Healer are each under-represented by 15 players.

### summary
In this task, I checked whether the observed player choices match the expected equal distribution across the four classes. I first compared the observed counts with the expected counts under the assumption that each class should have 25% of players, then used the chi-square goodness-of-fit test to see if the differences are statistically significant.

# Task 2: Chi-Square Test of Independence


In [7]:
# step 1
table = pd.DataFrame(
    [[90, 60, 30],
     [110, 140, 170]],
    index=["Churned", "Retained"],
    columns=["Free", "Basic", "Premium"]
)

table

,Free,Basic,Premium
Churned,90,60,30
Retained,110,140,170


In [8]:
# step 2
chi2_stat, p_value, df, expected = st.chi2_contingency(table)

In [9]:
#step 3
expected_df = pd.DataFrame(
    expected,
    index=table.index,
    columns=table.columns
)

expected_df

,Free,Basic,Premium
Churned,60.0,60.0,60.0
Retained,140.0,140.0,140.0


In [10]:
#step 4
print(f"Chi-square statistic: {chi2_stat:.2f}")
print(f"Degrees of freedom: {df}")
print(f"P-value: {p_value:.4f}")

Chi-square statistic: 42.86
Degrees of freedom: 2
P-value: 0.0000


In [11]:
#step 5
alpha = 0.05

if p_value < alpha:
    print("Reject H₀")
    print("There is evidence of an association between subscription tier and churn.")
else:
    print("Fail to reject H₀")
    print("There is not enough evidence of an association between subscription tier and churn.")

Reject H₀
There is evidence of an association between subscription tier and churn.


### summary 
In this task, I tested whether subscription tier and churn status are related. I created the contingency table, calculated the expected frequencies under the assumption of independence, and used the chi-square test to compare observed and expected values.

# Task 3: Compute Cramér's V


In [12]:
# step 1
def cramers_v(chi2, n, min_dim):
    return np.sqrt(chi2 / (n * min_dim))

In [13]:
#step 2
n = table.values.sum()
min_dim = min(table.shape) - 1

v = cramers_v(chi2_stat, n, min_dim)

print(f"Cramer's V: {v:.4f}")

Cramer's V: 0.2673


In [14]:
#step 3
if v < 0.10:
    strength = "weak"
elif v < 0.30:
    strength = "moderate"
elif v < 0.50:
    strength = "strong"
else:
    strength = "very strong"

print(f"Association strength: {strength}")

Association strength: moderate


### step 4
The relationship between subscription tier and churn is moderate because Cramér's V = 0.2673, which falls near the medium benchmark. This means subscription tier has a noticeable association with churn, but it is not strong enough to explain churn behavior alone. The marketing team should consider additional customer variables together with subscription tier.

### summary 
In this task, I calculated Cramér's V to measure how strong the relationship is after finding a significant chi-square result. This helps show whether the association between subscription tier and churn is weak, moderate, or strong in practical terms.